### DISCUSSION TITLE
Why does my custom draw function get ignored?

### DISCUSSION THREAD

**Discussion by AshfordN** [fltk-rs/fltk-rs#1045] [Q&A] [answered]
When I set a custom draw function for a `Frame` (and possibly other widgets), it seems as though the initial/default draw function is still called under certain circumstances. Below is an example that demonstrates this phenomenon.

This code....
```rust
use fltk::{prelude::*, *};

fn main() {
    let app_root = app::App::default();

    let mut win = window::Window::default()
        .with_size(300, 300)
        .center_screen()
        .with_label("Example");

    let mut frame0 = frame::Frame::default()
        .with_size(150, 20)
        .center_of_parent();
    frame0.set_frame(enums::FrameType::BorderBox);
    frame0.draw(|f| {
        let label = f.label();
        let txt = label.split(" ").nth(0).unwrap();
        let x = f.x();
        let y = f.y();
        let w = f.w();
        let h = f.h();

        draw::push_clip(x, y, w, h);
        draw::draw_box(f.frame(), x, y, w, h, f.color());
        draw::set_draw_color(f.label_color());
        draw::draw_text2(txt, x, y, w, h, f.align());
        draw::pop_clip();
    });
    frame0.set_label("text that extends beyond the frame");

    win.end();
    win.make_resizable(true);
    win.show();

    app_root.run().unwrap();
}
```

...produces this output
![output](https://user-images.githubusercontent.com/4706674/146136355-a40b8689-a6db-441d-ab08-8a325050f377.png)

I've also done more elaborate tests where I set the label after a timeout -- which seems to work -- but then deactivating the widget causes similar problems. Those additional examples were not included because I wanted to keep the example code as concise as possible and I'm assuming there's a single explanation for this behaviour.

**Comment by MoAlyousef** [comment, marked as answer]
The custom draw callback draws over the original draw method for any widget. It allows modification of the appearance not complete overriding. If you use a widget::Widget (which doesn't have an original draw method), you can in a sense completely override the draw method:
```rust
use fltk::{prelude::*, *};

fn main() {
    let app_root = app::App::default();

    let mut win = window::Window::default()
        .with_size(300, 300)
        .center_screen()
        .with_label("Example");

    let mut frame0 = widget::Widget::default()
        .with_size(150, 20)
        .center_of_parent();
    frame0.set_frame(enums::FrameType::BorderBox);
    frame0.draw(|f| {
        let label = f.label();
        let txt = label.split(" ").nth(0).unwrap();
        let x = f.x();
        let y = f.y();
        let w = f.w();
        let h = f.h();

        draw::push_clip(x, y, w, h);
        draw::draw_box(f.frame(), x, y, w, h, f.color());
        draw::set_draw_color(f.label_color());
        draw::set_font(enums::Font::Helvetica, 14);
        draw::draw_text2(txt, x, y, w, h, f.align());
        draw::pop_clip();
    });
    frame0.set_label("text that extends beyond the frame");

    win.end();
    win.make_resizable(true);
    win.show();

    app_root.run().unwrap();
}
```

![image](https://user-images.githubusercontent.com/37966791/146154770-6d0fff4d-e76b-43bd-8cd7-14acc68dca23.png)

**Reply by AshfordN** [reply]
Thanks for the quick and helpful reply as usual.

**Reply by AshfordN** [reply]
Why does the docs for `widget::Widget` stated that it shouldn't be instantiated in user code though? Is this because it doesn't have a default draw function, as you said?

**Reply by MoAlyousef** [reply]
Yes. But looking at it now, the wording isn't accurate. The docs for widget::Widget need improving.
